### PR TITLE
Add min cut partitioner for AOT+nvFuser

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Here we mark most of `torch.ops.nvprims` as something that can be recomputed in the backward passes (and hopefully fused).

TODO:
- [x] Add a test after https://github.com/pytorch/pytorch/pull/88186 is merged

cc @kevinstephano @jjsjann123 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/optimizations/training.py [hunk 1]
```diff
@@ -277,18 +277,62 @@ def prims_executor(gm, inputs, *, executor):
...
+    from torch.fx.experimental.proxy_tensor import make_fx
+
+    # AOT Autograd expects arguments of the traced function to be named exactly
+    # "primals, tangents"
+    def func(primals, tangents):
+        return joint_module(primals, tangents)
+
+    # First we trace the graph conditionally decomposing nodes
+    # that can be sent to the nvfuser executor
+    with TorchRefsNvfuserCapabilityMode():
```
Comment (jjsjann123) [inline torch/_dynamo/optimizations/training.py:296]: Does this mean that later we are speculatively lowering the partitioned graph to nvprim again in `prims_executor`?

should we just skip the speculative lowering, or is the second lowering there to catch some other decomposed op?

File: torch/_dynamo/optimizations/training.py [hunk 2]
```diff
@@ -277,18 +277,62 @@ def prims_executor(gm, inputs, *, executor):
...
+        prim_gm = make_fx(func)(*joint_inputs)
+
+    # all nvprims for now
+    recomputable_ops = {
```
Comment (jjsjann123) [inline torch/_dynamo/optimizations/training.py:300]: Wondering if we should define this `recomputable_ops` inside `nvfuser_prims.py`, where new nvprims are added? Just to avoid accidentally adding more normalization/reduction ops and having them default in recompute.

**Review by jjsjann123 — APPROVED** [review]
LGTM, minor comments

File: torch/_dynamo/optimizations/training.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_dynamo/optimizations/training.py:296]: Unfortunately yes.

Initially, I did leave this only in the partitioner code, but unfortunately, AOT has two different code paths for the case when at least one of the inputs requires grad and none requires grad, for the latter case "partition_fn" is not used at all, and there's no way to pass that information further without modifying AOT code. All inputs to the "fw_compiler" functions are detached and do not require grad, so it's not possible to determine this from within there.

File: torch/_dynamo/optimizations/training.py (hunk 2 above)
Comment (IvanYashchuk) [reply torch/_dynamo/optimizations/training.py:300]: Or maybe add compulsory markers to nvprims: this function is a reduction, and this one is a normalization, and so on.

File: torch/_dynamo/optimizations/training.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_dynamo/optimizations/training.py:299]: Okay with a bit of monkey patching unnecessary lowering step is avoided in https://github.com/pytorch/pytorch/pull/88204/commits/2a1103c918db8a451dbfe5e4c2981358fa17ed73

File: torch/_dynamo/optimizations/training.py (hunk 2 above)
Comment (jjsjann123) [reply torch/_dynamo/optimizations/training.py:303]: That sounds like a better idea. So let's put a TODO here and we can clean it up afterwards.

File: torch/_dynamo/optimizations/training.py (hunk 1 above)
Comment (jjsjann123) [reply torch/_dynamo/optimizations/training.py:299]: Looks good~ thx for patching this~

Comment (jjsjann123) [reply torch/_dynamo/optimizations/training.py:299]: A quick follow up question.

So now we are not lowering to nvprims after aot_autograd, does this mean `post-autograd` decomposition won't be lowered to nvprims now :exploding_head:

Comment (jjsjann123) [reply torch/_dynamo/optimizations/training.py:299]: I don't know exactly when `post-autograd` decomposition happens.. Hopefully it's not the case and we can still use it.
Can we get a CI test to guard/verify that behavior?

Comment (IvanYashchuk) [reply torch/_dynamo/optimizations/training.py:299]: We don't use decompositions (`aot_module_simplified(decompositions=None)`). Even if we used it, it would happen before calling the partitioning function.

**Comment by IvanYashchuk** [comment]
@jansel can you please approve to help merge changes to Dynamo's `nvprims_nvfuser` backend?

**Review by jansel — APPROVED** [review]
